pyuvdata/calfits.py

```diff
@@ -0,0 +1,194 @@
+                                      # self.flag_array[:, :, :, :, np.newaxis],
+                                      self.quality_array[:, :, :, :, np.newaxis]],
+                                     axis=-1)
+
+        if self.cal_type == 'delay':
+            # Set header variable for gain.
+            prihdr['CTYPE4'] = ('FREQS', 'Valid frequencies to apply delay.')
+            prihdr['CUNIT4'] = ('Hz', 'Units of frequecy.')
+            prihdr['CRVAL4'] = self.freq_array[0][0]
+            prihdr['CDELT4'] = self.channel_width
+            # set the last axis for number of arrays.
+            prihdr['CTYPE1'] = ('Narrays', 'Number of image arrays.')
+            prihdr['CUNIT1'] = ('Integer', 'Number of image arrays. Value.')
+            prihdr['CRVAL1'] = (2, 'Number of image arrays.')
+            prihdr['CDELT1'] = 1
+
+            pridata = np.concatenate([self.delay_array[:, :, :, :, np.newaxis],
+                                      # self.flag_array[:, :, :, :, np.newaxis],
+                                      self.quality_array[:, :, :, :, np.newaxis]],
+                                     axis=-1)
+
+        if self.cal_type == 'unknown':
+            raise ValueError("unknown calibration type. Do not know how to"
+                             "store parameters")
+
+        # header ctypes for NAXIS
+        prihdr['CTYPE5'] = ('ANTS', 'Antenna numbering.')
+        prihdr['CUNIT5'] = 'Integer'
+        prihdr['CRVAL5'] = 0
+        prihdr['CDELT5'] = -1
+
+        prihdr['CTYPE3'] = ('TIME', 'Time axis.')
+        prihdr['CUNIT3'] = ('JD', 'Time in julian date format')
+        prihdr['CRVAL3'] = self.time_array[0]
+        prihdr['CDELT3'] = self.integration_time
+
+        # more checks for polarization. check ordering.
+        prihdr['CTYPE2'] = ('JONES', 'Jones matrix array')
+        prihdr['CUNIT2'] = ('Integer', 'representative integer for polarization.')
+        prihdr['CRVAL2'] = self.jones_array[0]  # always start with first jones.
+        prihdr['CDELT2'] = -1
+
+        prihdu = fits.PrimaryHDU(data=pridata, header=prihdr)
+
+        col1 = fits.Column(name='ANTNAME', format='8A',
+                           array=self.antenna_names)
+        col2 = fits.Column(name='ANTINDEX', format='D',
+                           array=self.antenna_numbers)
+        cols = fits.ColDefs([col1, col2])
+        ant_hdu = fits.BinTableHDU.from_columns(cols)
+
+        sechdr = fits.Header()
+        secdata = self.flag_array.astype(np.int64) # Can't be bool
+        prihdu = fits.PrimaryHDU(data=pridata, header=prihdr)
+        sechdu = fits.ImageHDU(data=secdata, header=sechdr)
+        hdulist = fits.HDUList([prihdu, ant_hdu, sechdu])
+        hdulist.writeto(filename, clobber=clobber)
+
+    def read_calfits(self, filename, run_check=True, run_check_acceptability=True):
+        F = fits.open(filename)
+        data = F[0].data
+        hdr = F[0].header.copy()
+
+        antdata = F[1].data
```

Comment:
You should be checking for the HDU name here, rather than hard coding an ordering. Fits files are explicitly allowed to have any ordering of HDUs after the first one. Check out line 223 in uvfits.py to see how we find the antenna table.